synapse/handlers/sso.py

```diff
@@ -94,3 +108,115 @@ async def get_sso_user_by_remote_user_id(
 
         # No match.
         return None
+
+    async def get_mxid_from_sso(
+        self,
+        auth_provider_id: str,
+        remote_user_id: str,
+        user_agent: str,
+        ip_address: str,
+        sso_to_matrix_id_mapper: Callable[[int], Awaitable[UserAttributes]],
+        allow_existing_users: bool = False,
+    ):
+        """
+
+        Args:
+            auth_provider_id: A unique identifier for this SSO provider, e.g.
+                "oidc" or "saml".
+            remote_user_id: The unique identifier from the SSO provider.
+            user_agent: The user agent of the client making the request.
+            ip_address: The IP address of the client making the request.
+            sso_to_matrix_id_mapper: A callable to generate the user attributes.
+                The only parameter is an integer which represents the amount of
+                times the returned mxid localpart mapping has failed.
+            allow_existing_users: True if the localpart returned from the
+                mapping provider can be linked to an existing matrix ID.
+
+        Returns:
+             The user ID associated with the SSO response.
+
+        Raises:
+            MappingException if there was a problem mapping the response to a user.
+            RedirectException: some mapping providers may raise this if they need
+                to redirect to an interstitial page.
+
+        """
+        # first of all, check if we already have a mapping for this user
+        previously_registered_user_id = await self.get_sso_user_by_remote_user_id(
+            auth_provider_id, remote_user_id,
+        )
+        if previously_registered_user_id:
+            return previously_registered_user_id
+
+        # Otherwise, generate a new user.
+        for i in range(self._MAP_USERNAME_RETRIES):
+            try:
+                attributes = await sso_to_matrix_id_mapper(i)
+            except Exception as e:
+                raise MappingException(
+                    "Could not extract user attributes from SSO response: " + str(e)
+                )
+
+            logger.debug(
+                "Retrieved user attributes from user mapping provider: %r (attempt %d)",
+                attributes,
+                i,
+            )
+
+            if not attributes.localpart:
+                raise MappingException(
+                    "Error parsing SSO response: SSO mapping provider plugin "
+                    "did not return a localpart value"
+                )
+
+            # Check if this mxid already exists
+            user_id = UserID(attributes.localpart, self.server_name).to_string()
+            users = await self.store.get_users_by_id_case_insensitive(user_id)
+            if users and allow_existing_users:
```

Comment:
I think that this means that if `allow_existing_users` is true we'll either match on the first iteration or error. I couldn't come up with a situation where we wouldn't do this.

Reply:
Do we even want to retry if `allow_existing_users` is set to true?

Reply:
I had the same debate internally and initially had the same though, but then convinced myself it didn't matter. I believe what my above comment is trying to say is that no matter what this will end on the first loop if `allow_existing_users` is true:

1. If `users` is empty we've found an unused ID and can claim it (the `elif not users` branch).
2. If `users` is truthy, we will either use one of those IDs (if there's only one, or an exact case match) or raise an error (the `MappingException` branch).

So I think the answer is "yes we should not retry", but we essentially already do that.

Reply:
I added some of this info to the docstring for this method too.

Reply:
Ok, I think I see. Can you just add a quick comment here to say something like "Note, if `allow_existing_users` is true we won't loop"?

Reply:
Will do!

Reply:
See 6d11ff81607bb5f9ad33b4a760ecb4cf55380765